src/neural/blas/winograd_convolution3.h

```diff
@@ -0,0 +1,84 @@
+/*
+ This file is part of Leela Chess Zero.
+ Copyright (C) 2018 The LCZero Authors
+
+ Leela Chess is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ Leela Chess is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <vector>
+
+namespace lczero {
+
+// Convolution 3x3 on a 8x8 board using the Winograd algorithm.
+//
+// Ref:
+//
+// Fast Algorithms for Convolutional Neural Networks
+// https://arxiv.org/abs/1509.09308
+//
+// https://ai.intel.com/winograd/
+// https://ai.intel.com/winograd-2/
+
+// Convolution 3x3 using the Winograd algorithm
+class WinogradConvolution3 {
+ public:
+  // Create the zero-padded U matrix
+  static std::vector<float> ZeropadU(const std::vector<float>& U,
+                                     const size_t outputs,
+                                     const size_t channels,
+                                     const size_t outputs_pad,
+                                     const size_t channels_pad);
+
+  // Transform the weights
+  static std::vector<float> TransformF(const std::vector<float>& f,
+                                       const size_t outputs,
+                                       const size_t channels);
+
+  // Allocate for the largest batch size
```

Comment:
Allocate what? And why is function called "WinograteConvolution" and not "Allocate<Something>"?

Reply:
It's a class that does the forward inference of a convolution 3x3 layer.
The matrices M and U are memory resources that are allocated once for all for the duration of the computation. The batch size of the computation may be larger than the actual largest internal batch.  M and U are allocated for the largest batch size.

So the class name is WinogradConvolution3. But the comment can be enhanced.

Reply:
I settled for 
```
  // The instance will allocate memory resources for the
  // largest batch size, and the largest input and output
  // layers.
```